database/dtb.ksy

```diff
@@ -0,0 +1,152 @@
+meta:
+  id: dtb
+  title: Flattened Devicetree Format
+  file-extension: dtb
+  application:
+    - Linux
+    - Das U-Boot
+  xref:
+    wikidata: Q16960371
+  tags:
+    - linux
+  license: CC0-1.0
+  ks-version: 0.9
+  encoding: ASCII
+  endian: be
+doc: |
+  Also referred to as Devicetree Blob (DTB). It is a flat
+  binary encoding of data (primarily devicetree data, although
+  other data is possible as well).
+
+  On Linux systems that support this the blobs can be accessed in
+  `/sys/firmware/fdt`:
+
+  - <https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-firmware-ofw>
+
+  The encoding of strings used in the strings block and struct block is
+  actually a subset of ASCII (section "2.2.1 Node Names"):
+
+  <https://github.com/devicetree-org/devicetree-specification/releases/download/v0.3/devicetree-specification-v0.3.pdf>
+
+  Example files:
+
+  - <https://github.com/qemu/qemu/tree/master/pc-bios>
+doc-ref:
+  - https://github.com/devicetree-org/devicetree-specification/releases/download/v0.3/devicetree-specification-v0.3.pdf
+    Flattened Devicetree (DTB) Format
+  - https://elinux.org/images/f/f4/Elc2013_Fernandes.pdf
+seq:
+  - id: magic
+    -orig-id: magic
+    contents: [0xd0, 0x0d, 0xfe, 0xed]
+  - id: total_size
+    -orig-id: totalsize
+    type: u4
+  - id: ofs_structure_block
+    -orig-id: off_dt_struct
+    type: u4
+  - id: ofs_strings_block
+    -orig-id: off_dt_strings
+    type: u4
+  - id: ofs_memory_reservation_block
+    -orig-id: off_mem_rsvmap
+    type: u4
+  - id: version
+    type: u4
+  - id: last_compatible_version
```

Comment:
This name is misleading, at least to me - if I see **last** compatible version I imagine the latest one (i.e. the one that came _last_), not the lowest or earliest. However, it's clear that the [spec](https://github.com/devicetree-org/devicetree-specification/releases/download/v0.3/devicetree-specification-v0.3.pdf) intended the exact opposite:

> <dl>
> <dt>last_comp_version</dt>
> <dd>This field shall contain the lowest version of the devicetree data structure with which the version
> used is backwards compatible. So, for the structure as defined in this document (version 17), this field shall contain
> 16 because version 17 is backwards compatible with version 16, but not earlier versions. As per section 5.1, a
> DTSpec boot program should provide a devicetree in a format which is backwards compatible with version 16, and
> thus this field shall always contain 16.</dd>

I'm not sure why did the spec authors choose `last_` adjective in the name. Probably when you imagine the latest version as the "first", then you go further into the version history and declare the oldest compatible version as the "last" that would the current version be compatible with (if we went beyond the "last", any older versions would no longer work). However, this concept seems to be upside down - I guess it's natural that "last" should mean latest (the last one that has been released so far) and the "first" would mean the initial version, i.e. the oldest version. In this justification for the spec's naming that I came up with, I needed to reverse the timeline and the role of "first" and "last" pointers (we go _back in time_).

Does `first_compatible_version`, `oldest_compatible_version`, `lowest_compatible_version` or `min_compatible_version` make more sense? Which one would you choose?